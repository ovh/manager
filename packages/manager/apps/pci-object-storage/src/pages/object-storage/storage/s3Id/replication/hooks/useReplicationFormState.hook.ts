import { UseFormReturn } from 'react-hook-form';
import { ReplicationFormValues } from '../schemas/replicationForm.schema';

export const useReplicationFormState = (
  form: UseFormReturn<ReplicationFormValues>,
) => {
  const formValues = form.watch();

  const hasTags =
    formValues.tags &&
    Object.values(formValues.tags).some(
      (tag) => tag.key.trim() !== '' || tag.value.trim() !== '',
    );
  const isDeleteMarkerEnabled =
    formValues.deleteMarkerReplication === 'enabled';

  return {
    showScopeFields: formValues.isReplicationApplicationLimited,
    showStorageClassField: formValues.useStorageClass,
    isDeleteMarkerDisabled: hasTags,
    isTagsDisabled: isDeleteMarkerEnabled,
  };
};
