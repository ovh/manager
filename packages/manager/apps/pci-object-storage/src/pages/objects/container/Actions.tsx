import { createSearchParams, useParams, useNavigate } from 'react-router-dom';
import { ActionMenu, useNotifications } from '@ovh-ux/manager-react-components';
import { Translation, useTranslation } from 'react-i18next';
import { ApiError } from '@ovh-ux/manager-core-api';

import { useMemo } from 'react';
import { TStorage } from '@/api/data/storages';
import { OBJECT_CONTAINER_MODE_LOCAL_ZONE } from '@/constants';
import { isSwiftType } from '@/helpers';
import { useUpdateStorageType } from '@/api/hooks/useStorages';

const PUBLIC_STORAGE_TYPES = ['public', 'static'];

export function Actions({
  storage,
  isEmptyUsers,
}: Readonly<{ storage: TStorage; isEmptyUsers: boolean }>) {
  const { t } = useTranslation('containers');

  const navigate = useNavigate();
  const { projectId } = useParams();
  const { addSuccess, addError, clearNotifications } = useNotifications();

  const { isPending, updateStorageType } = useUpdateStorageType({
    projectId,
    onSuccess: (containerType) => {
      addSuccess(
        <Translation ns="containers">
          {(_t) =>
            _t(
              `pci_projects_project_storages_containers_toggle_${containerType}_succeed`,
              {
                name: storage.name,
              },
            )
          }
        </Translation>,
        true,
      );
    },
    onError(error: ApiError) {
      addError(
        <Translation ns="containers">
          {(_t) =>
            _t('pci_projects_project_storages_containers_toggle_fail', {
              message: error?.response?.data?.message || error?.message || null,
            })
          }
        </Translation>,
        true,
      );
    },
  });

  const uid = useMemo(
    () =>
      `${storage.name}-${storage.deploymentMode}-${storage.region}${
        storage.s3StorageType ? '-s3' : ''
      }`.replace(/\./g, '-'),
    [storage],
  );

  const items = [
    {
      id: 0,
      label: t('pci_projects_project_storages_containers_view_add_user_label'),
      onClick: () =>
        isEmptyUsers
          ? navigate('./emptyUser')
          : navigate({
              pathname: `./addUser`,
              search: `?${createSearchParams({
                containerId: storage.name,
              })}`,
            }),
      condition:
        storage.s3StorageType &&
        (!storage.deploymentMode ||
          storage.deploymentMode !== OBJECT_CONTAINER_MODE_LOCAL_ZONE),
    },
    {
      id: 1,
      label: t('pci_projects_project_storages_containers_view_object_label'),
      onClick: () =>
        navigate({
          pathname: `./${storage.name}`,
          search: `?${createSearchParams({
            region: storage.region,
          })}`,
        }),
    },
    {
      id: 2,
      onClick: () => {
        clearNotifications();
        updateStorageType(
          storage.id,
          PUBLIC_STORAGE_TYPES.includes(storage.containerType)
            ? 'private'
            : 'public',
        );
      },
      label: t(
        PUBLIC_STORAGE_TYPES.includes(storage.containerType)
          ? 'pci_projects_project_storages_containers_switch_to_private_label'
          : 'pci_projects_project_storages_containers_switch_to_public_label',
      ),
      condition: isSwiftType(storage),
      isDisabled: isPending,
    },
    {
      id: 3,
      label: t('pci_projects_project_storages_containers_delete_label'),
      onClick: () =>
        navigate({
          pathname: `./delete`,
          search: `?${createSearchParams({
            containerId: storage.name,
            region: storage.region,
          })}`,
        }),
    },
  ]
    .filter((i) => !('condition' in i) || i.condition)
    .map((i, index) => ({
      id: index,
      ...i,
    }));

  return <ActionMenu id={uid} items={items} isCompact />;
}
