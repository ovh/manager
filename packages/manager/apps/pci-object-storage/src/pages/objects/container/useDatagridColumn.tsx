import {
  DataGridTextCell,
  DatagridColumn,
} from '@ovh-ux/manager-react-components';
import { createSearchParams, useHref, useParams } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import { useBytes } from '@ovh-ux/manager-pci-common';
import { OdsBadge, OdsLink } from '@ovhcloud/ods-components/react';
import { TStorage } from '@/api/data/storages';
import {
  OBJECT_CONTAINER_MODE_LOCAL_ZONE,
  OBJECT_CONTAINER_MODE_MONO_ZONE,
  OBJECT_CONTAINER_MODE_MULTI_ZONES,
} from '@/constants';
import { Actions } from './Actions';
import { useUsers } from '@/api/hooks/useUser';

export const useDatagridColumn = () => {
  const { i18n, t } = useTranslation(['containers', 'pci-common']);

  const { formatBytes } = useBytes();
  const { projectId } = useParams();

  const { validUsersWithCredentials } = useUsers(projectId);

  const columns: DatagridColumn<TStorage>[] = [
    {
      id: 'name',
      cell: (props: TStorage) => (
        <DataGridTextCell>
          <OdsLink
            color="primary"
            href={useHref({
              pathname: `./${props.id || props.name}`,
              search: `?${createSearchParams({
                region: props.region,
              })}`,
            })}
            label={props.name}
          />
        </DataGridTextCell>
      ),
      label: t('pci_projects_project_storages_containers_name_label'),
    },
    {
      id: 'region',
      cell: (props: TStorage) => (
        <DataGridTextCell>{props.region}</DataGridTextCell>
      ),
      label: t('pci_projects_project_storages_containers_region_label'),
    },
    {
      id: 'mode',
      cell: (props: TStorage) => {
        if (props.deploymentMode === OBJECT_CONTAINER_MODE_MULTI_ZONES) {
          return (
            <DataGridTextCell>
              <OdsBadge
                color="information"
                size="sm"
                label={props.mode}
                className="badge-deployment-multizone"
              />
            </DataGridTextCell>
          );
        }
        if (props.deploymentMode === OBJECT_CONTAINER_MODE_MONO_ZONE) {
          return (
            <DataGridTextCell>
              <OdsBadge color="information" size="sm" label={props.mode} />
            </DataGridTextCell>
          );
        }
        if (props.deploymentMode === OBJECT_CONTAINER_MODE_LOCAL_ZONE) {
          return (
            <DataGridTextCell>
              <OdsBadge color="promotion" size="sm" label={props.mode} />
            </DataGridTextCell>
          );
        }
        return undefined;
      },
      label: t(
        'pci_projects_project_storages_containers_deployment_mode_label',
      ),
    },
    {
      id: 'offer',
      cell: (props: TStorage) => (
        <DataGridTextCell>{props.offer}</DataGridTextCell>
      ),
      label: t('pci_projects_project_storages_containers_offer_label'),
    },
    {
      id: 'containerCount',
      cell: (props: TStorage) => (
        <DataGridTextCell>
          {props.deploymentMode === OBJECT_CONTAINER_MODE_LOCAL_ZONE
            ? 'n/a'
            : props.containerCount}
        </DataGridTextCell>
      ),
      label: t('pci_projects_project_storages_containers_storedObjects_label'),
    },
    {
      id: 'usedSpace',
      cell: (props: TStorage) => {
        if (props.usedSpace > 0) {
          return (
            <DataGridTextCell>
              {formatBytes(props.usedSpace, 2, 1024)}
            </DataGridTextCell>
          );
        }
        if (props.deploymentMode === OBJECT_CONTAINER_MODE_LOCAL_ZONE) {
          return <DataGridTextCell>n/a</DataGridTextCell>;
        }
        return <DataGridTextCell>-</DataGridTextCell>;
      },
      label: t('pci_projects_project_storages_containers_storedBytes_label'),
    },
    {
      id: 'containerType',
      cell: (props: TStorage) => {
        if (
          i18n.exists(
            `containers:pci_projects_project_storages_containers_containerType_${props.containerType}`,
          )
        ) {
          return (
            <DataGridTextCell>
              {t(
                `pci_projects_project_storages_containers_containerType_${props.containerType}`,
              )}
            </DataGridTextCell>
          );
        }
        return undefined;
      },
      label: t('pci_projects_project_storages_containers_containerType_label'),
    },
    {
      id: 'actions',
      cell: (props: TStorage) => (
        <Actions
          storage={props}
          isEmptyUsers={validUsersWithCredentials?.length === 0}
        />
      ),
      label: '',
      isSortable: false,
    },
  ];

  return columns;
};
