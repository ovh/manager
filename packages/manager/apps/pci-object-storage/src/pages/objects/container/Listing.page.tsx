import {
  Datagrid,
  FilterAdd,
  FilterList,
  Notifications,
  RedirectionGuard,
  useColumnFilters,
  useDataGrid,
  useFeatureAvailability,
  useNotifications,
} from '@ovh-ux/manager-react-components';
import { ShellContext } from '@ovh-ux/manager-react-shell-client';
import { Suspense, useContext, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Outlet, useNavigate, useParams } from 'react-router-dom';
import { FilterCategories, FilterComparator } from '@ovh-ux/manager-core-api';
import {
  OdsButton,
  OdsInput,
  OdsPopover,
  OdsSpinner,
} from '@ovhcloud/ods-components/react';
import { useStorages } from '@/api/hooks/useStorages';
import { useDatagridColumn } from './useDatagridColumn';
import { AVAILABILITY } from '@/constants';

export default function ListingPage() {
  const { t } = useTranslation([
    'pci-storages-containers',
    'containers/add',
    'pci-common',
  ]);
  const { projectId } = useParams();
  const navigate = useNavigate();
  const { tracking } = useContext(ShellContext).shell;
  const { clearNotifications } = useNotifications();
  const columns = useDatagridColumn();
  const { pagination, setPagination, sorting, setSorting } = useDataGrid({
    id: 'name',
    desc: false,
  });
  const { filters, addFilter, removeFilter } = useColumnFilters();
  const [searchField, setSearchField] = useState('');

  const {
    data: availability,
    isPending: isAvailabilityPending,
  } = useFeatureAvailability([AVAILABILITY.LOCALZONE, AVAILABILITY['3AZ']]);

  const {
    allStorages,
    paginatedStorages,
    isRefreshing,
    isPending: isStoragesPending,
    refresh,
  } = useStorages(projectId, pagination, sorting, filters, {
    isLocalZoneAvailable: availability?.[AVAILABILITY.LOCALZONE],
    is3azAvailable: availability?.[AVAILABILITY['3AZ']],
  });

  const isPending = isAvailabilityPending || isStoragesPending;

  if (isPending) {
    return <OdsSpinner size="md" />;
  }

  return (
    <RedirectionGuard
      isLoading={isPending}
      condition={!isPending && allStorages?.resources.length === 0}
      route="./onboarding"
    >
      <Notifications />
      <div className="sm:flex items-center justify-between mt-4">
        <OdsButton
          size="sm"
          color="primary"
          className="xs:mb-0.5 sm:mb-0"
          icon="plus"
          label={t(
            'containers/add:pci_projects_project_storages_containers_add_title',
          )}
          onClick={() => {
            clearNotifications();
            navigate('./new');
            tracking?.trackClick({
              name: 'PCI_PROJECTS_CONTAINERS_ADD',
              type: 'navigaton',
            });
          }}
        />
        <div className="justify-between flex space-x-4">
          <OdsButton
            size="sm"
            variant="outline"
            color="primary"
            className="xs:mb-0.5 sm:mb-0"
            isDisabled={isRefreshing || undefined}
            onClick={refresh}
            icon="refresh"
            label=""
          />
          <OdsInput
            type="search"
            name="searchField"
            className="w-[70%]"
            value={searchField}
            onKeyDown={(event) => {
              if (event.key === 'Enter') {
                setPagination({
                  pageIndex: 0,
                  pageSize: pagination.pageSize,
                });
                addFilter({
                  key: 'name',
                  value: searchField,
                  comparator: FilterComparator.Includes,
                  label: '',
                });
                setSearchField('');
              }
            }}
            onOdsChange={(event) => {
              const value = event.detail.value.toString();
              setSearchField(value);
            }}
          />
          <div>
            <OdsButton
              slot="popover-trigger"
              id="popover-filter"
              size="sm"
              color="primary"
              label={t('pci-common:common_criteria_adder_filter_label')}
              variant="outline"
              icon="filter"
            />
          </div>
          <OdsPopover triggerId="popover-filter">
            <FilterAdd
              columns={[
                {
                  id: 'name',
                  label: t(
                    'pci_projects_project_storages_containers_name_label',
                  ),
                  comparators: FilterCategories.String,
                },
                {
                  id: 'region',
                  label: t(
                    'pci_projects_project_storages_containers_region_label',
                  ),
                  comparators: FilterCategories.String,
                },
                {
                  id: 'mode',
                  label: t(
                    'pci_projects_project_storages_containers_deployment_mode_label',
                  ),
                  comparators: FilterCategories.String,
                },
                {
                  id: 'offer',
                  label: t(
                    'pci_projects_project_storages_containers_offer_label',
                  ),
                  comparators: FilterCategories.String,
                },
                {
                  id: 'storedObjects',
                  label: t(
                    'pci_projects_project_storages_containers_storedObjects_label',
                  ),
                  comparators: FilterCategories.Numeric,
                },
              ]}
              onAddFilter={(addedFilter, column) => {
                setPagination({
                  pageIndex: 0,
                  pageSize: pagination.pageSize,
                });
                addFilter({
                  ...addedFilter,
                  label: column.label,
                });
              }}
            />
          </OdsPopover>
        </div>
      </div>

      {filters?.length > 0 && (
        <div className="my-5">
          <FilterList filters={filters} onRemoveFilter={removeFilter} />
        </div>
      )}

      <div className="mt-8">
        <Datagrid
          columns={columns}
          items={paginatedStorages?.rows || []}
          totalItems={allStorages?.resources?.length || 0}
          pagination={pagination}
          onPaginationChange={setPagination}
          sorting={sorting}
          onSortChange={setSorting}
          className="overflow-x-visible"
        />
      </div>

      <Suspense>
        <Outlet />
      </Suspense>
    </RedirectionGuard>
  );
}
