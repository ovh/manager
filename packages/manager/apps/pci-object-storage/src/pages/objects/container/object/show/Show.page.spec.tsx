import { UseQueryResult } from '@tanstack/react-query';
import { describe, vi } from 'vitest';
import { render } from '@testing-library/react';
import { useProductMaintenance } from '@ovh-ux/manager-react-components';
import { TServerContainer } from '@/api/data/container';
import {
  usePaginatedObjects,
  useServerContainer,
} from '@/api/hooks/useContainer';
import ShowPage from './Show.page';
import { wrapper } from '@/wrapperRenders';
import { useStorageEndpoint } from '@/api/hooks/useStorages';

vi.mock('@/api/hooks/useContainer', () => {
  return {
    useServerContainer: vi.fn(),
    usePaginatedObjects: vi.fn(),
  };
});

vi.mock('@/api/hooks/useStorages', async () => {
  const mod = await vi.importActual('@/api/hooks/useStorages');
  return {
    ...mod,
    useStorageEndpoint: vi.fn(),
  };
});

vi.mock('date-fns', () => {
  const formatMock = vi.fn(() => 'formattedDate');
  return {
    format: formatMock,
  };
});

describe('Show page', () => {
  it('should display spinner', () => {
    vi.fn(useProductMaintenance).mockReturnValue({
      hasMaintenance: false,
    } as never);
    vi.mocked(useServerContainer).mockReturnValue({
      isPending: true,
      data: {},
    } as UseQueryResult<TServerContainer>);
    vi.mocked(usePaginatedObjects).mockReturnValue({
      isPending: true,
    } as never);
    vi.mocked(useStorageEndpoint).mockReturnValue({
      isPending: true,
    } as never);
    const { container } = render(<ShowPage />, { wrapper });
    expect(container).toMatchSnapshot();
  });

  it('should display datagrid', () => {
    const serverContainer = ({
      name: 'name',
      id: 'id',
      region: 'region',
      objects: [
        {
          name: 'name',
          size: 1,
          lastModified: new Date('2023-10-31T23:00:00.000Z'),
          contentType: 'contentType',
        },
      ],
    } as unknown) as TServerContainer;
    vi.mocked(useServerContainer).mockReturnValue({
      data: serverContainer,
      isPending: false,
    } as UseQueryResult<TServerContainer>);
    vi.mocked(usePaginatedObjects).mockReturnValue({
      isPending: false,
      paginatedObjects: {
        rows: serverContainer.objects,
        totalRows: 1,
      },
    } as never);
    vi.fn(useProductMaintenance).mockReturnValue({
      hasMaintenance: false,
    } as never);
    vi.mocked(useStorageEndpoint).mockReturnValue({
      url: 'https://api.ovh.com',
      isPending: false,
    } as never);
    const { container } = render(<ShowPage />, { wrapper });
    expect(container).toMatchSnapshot();
  });
});
