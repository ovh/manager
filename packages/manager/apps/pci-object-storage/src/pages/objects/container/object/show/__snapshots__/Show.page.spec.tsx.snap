// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Show page > should display datagrid 1`] = `
<div>
  <div
    class="py-8 px-4 md:py-9 md:px-10 md:mt-2"
  >
    <div>
      <ods-breadcrumb>
        <ods-breadcrumb-item
          href="mockProjectUrl"
        />
        <ods-breadcrumb-item
          label="pci_projects_project_storages_containers_object_title"
        />
        <ods-breadcrumb-item
          href=""
          label="name"
        />
      </ods-breadcrumb>
    </div>
    <div
      class="mt-[24px]"
    >
      <div
        class="flex items-center justify-between"
      >
        <div>
          <div
            class=" m-0 p-0 text-[--ods-color-heading] text-[36px] leading-[45px] font-bold mb-[24px]"
          >
            name
          </div>
        </div>
      </div>
    </div>
    <div
      class="mb-6"
    />
    <div
      data-testid="notifications"
    />
    <div
      class="grid grid-cols-12 gap-4 border border-solid border-[#bef1ff] bg-[#f5feff] rounded-md mt-6 py-8 px-12"
    >
      <div
        class="grid gap-2 col-span-12 md:col-span-4"
      >
        <div
          class="mb-4"
        >
          <ods-text>
            <span>
              pci_projects_project_storages_containers_container_region
            </span>
          </ods-text>
        </div>
        <div
          class="mb-4"
        >
          <ods-text>
            <span>
              pci_projects_project_storages_containers_container_object_info_storedObjects
            </span>
          </ods-text>
        </div>
        <div
          class="mb-4"
        >
          <ods-text>
            <span>
              pci_projects_project_storages_containers_container_info_storedBytes
            </span>
          </ods-text>
        </div>
      </div>
      <div
        class="grid col-span-12 md:col-span-8 gap-4"
      >
        <ods-form-field>
          <div
            class="flex gap-2 items-center"
            slot="label"
          >
            <ods-text
              class="font-bold-class"
              preset="caption"
            >
              pci_projects_project_storages_containers_container_info_id
            </ods-text>
          </div>
          <ods-clipboard
            class="w-[100%]"
            data-testid="clipboard"
            label-copy="clipboard_copy"
            label-copy-success="clipboard_copy_success"
            value="id"
          />
        </ods-form-field>
        <ods-form-field>
          <div
            class="flex gap-2 items-center"
            slot="label"
          >
            <ods-text
              class="font-bold-class"
              preset="caption"
            >
              pci_projects_project_storages_containers_container_info_publicUrl
            </ods-text>
            <div>
              <ods-icon
                class="cursor-help"
                id="public-url-popover"
                name="circle-question"
              />
              <ods-popover
                trigger-id="public-url-popover"
              >
                <ods-text
                  preset="caption"
                >
                  pci_projects_project_storages_containers_container_info_publicUrl_help
                </ods-text>
              </ods-popover>
            </div>
          </div>
          <ods-clipboard
            class="w-[100%]"
            data-testid="clipboard"
            label-copy="clipboard_copy"
            label-copy-success="clipboard_copy_success"
            value="https://api.ovh.com"
          />
        </ods-form-field>
        <ods-form-field>
          <div
            class="flex gap-2 items-center"
            slot="label"
          >
            <ods-text
              class="font-bold-class"
              preset="caption"
            >
              pci_projects_project_storages_containers_container_object_info_staticUrl
            </ods-text>
            <div>
              <ods-icon
                class="cursor-help"
                id="virtual-host-popover"
                name="circle-question"
              />
              <ods-popover
                trigger-id="virtual-host-popover"
              >
                <ods-text
                  preset="caption"
                >
                  pci_projects_project_storages_containers_container_object_info_staticUrl_help
                </ods-text>
              </ods-popover>
            </div>
          </div>
          <ods-clipboard
            class="w-[100%]"
            data-testid="clipboard"
            label-copy="clipboard_copy"
            label-copy-success="clipboard_copy_success"
          />
        </ods-form-field>
      </div>
    </div>
    <div
      class="sm:flex items-center justify-between mt-8"
    >
      <ods-button
        icon="plus"
        label="pci_projects_project_storages_containers_container_add_object_label"
        size="sm"
      />
      <div
        class="flex justify-center gap-4 ml-auto"
      >
         
        <ods-input
          class="min-w-[15rem]"
          name="searchField"
          value=""
        />
        <ods-button
          icon="magnifying-glass"
          label=""
          size="sm"
        />
        <ods-button
          color="primary"
          icon="filter"
          id="filterPopoverTrigger"
          label="common_criteria_adder_filter_label"
          size="sm"
          variant="outline"
        />
        <ods-popover
          trigger-id="filterPopoverTrigger"
        >
          <div>
            <ods-form-field
              class="w-full"
            >
              <div
                slot="label"
              >
                <span
                  class="text-[--ods-color-heading] leading-[22px]"
                >
                  common_criteria_adder_column_label
                </span>
              </div>
              <ods-select
                data-testid="add-filter_select_idColumn"
                name="add-filter_select_idColumn"
                value="name"
              >
                <option
                  value="name"
                >
                  pci_projects_project_storages_containers_container_name_label
                </option>
                <option
                  value="lastModified"
                >
                  pci_projects_project_storages_containers_container_lastModified_label
                </option>
                <option
                  value="storageClass"
                >
                  pci_projects_project_storages_containers_container_storage_class_label
                </option>
              </ods-select>
            </ods-form-field>
          </div>
          <div>
            <ods-form-field
              class="mt-2 w-full"
            >
              <div
                slot="label"
              >
                <span
                  class="text-[--ods-color-heading] leading-[22px]"
                >
                  common_criteria_adder_operator_label
                </span>
              </div>
              <div>
                <ods-select
                  name="add-operator-name"
                  value="includes"
                >
                  <option
                    value="includes"
                  >
                    common_criteria_adder_operator_includes
                  </option>
                  <option
                    value="starts_with"
                  >
                    common_criteria_adder_operator_starts_with
                  </option>
                  <option
                    value="ends_with"
                  >
                    common_criteria_adder_operator_ends_with
                  </option>
                  <option
                    value="is_equal"
                  >
                    common_criteria_adder_operator_is_equal
                  </option>
                  <option
                    value="is_different"
                  >
                    common_criteria_adder_operator_is_different
                  </option>
                </ods-select>
              </div>
            </ods-form-field>
          </div>
          <div>
            <ods-form-field
              class="mt-2 w-full"
            >
              <div
                slot="label"
              >
                <span
                  class="text-[--ods-color-heading] leading-[22px]"
                >
                  common_criteria_adder_value_label
                </span>
              </div>
              <ods-input
                class="border"
                data-testid="filter-add_value-input"
                name="filter-add_value-input"
                type="text"
                value=""
              />
            </ods-form-field>
          </div>
          <div>
            <ods-button
              class="mt-4 w-full filter-add-button-submit"
              data-testid="filter-add_submit"
              is-disabled="true"
              label="common_criteria_adder_submit_label"
              size="sm"
            />
          </div>
        </ods-popover>
      </div>
    </div>
    <div
      class="mt-8"
    />
    <div
      class="mt-8"
    >
      <div>
        <div
          class="contents px-[1px] overflow-x-visible"
        >
          <ods-table
            class="overflow-x-visible"
          >
            <!---->
            <table
              class="w-full border-collapse"
              style="--expander-column-width: 2.5rem;"
            >
              <thead>
                <tr>
                  <th
                    class="text-left pl-4 h-11 whitespace-nowrap "
                  >
                    <div
                      class="cursor-pointer select-none"
                      data-testid="header-name"
                    >
                      <span>
                        pci_projects_project_storages_containers_container_name_label
                      </span>
                      <span
                        class="align-middle inline-block h-4 -mt-6"
                      >
                        <ods-icon
                          class="invisible"
                          name="arrow-down"
                        />
                      </span>
                    </div>
                  </th>
                  <th
                    class="text-left pl-4 h-11 whitespace-nowrap "
                  >
                    <div
                      class="cursor-pointer select-none"
                      data-testid="header-lastModified"
                    >
                      <span>
                        pci_projects_project_storages_containers_container_lastModified_label
                      </span>
                      <span
                        class="align-middle inline-block h-4 -mt-6"
                      >
                        <ods-icon
                          class="invisible"
                          name="arrow-down"
                        />
                      </span>
                    </div>
                  </th>
                  <th
                    class="text-left pl-4 h-11 whitespace-nowrap "
                  >
                    <div
                      class="cursor-pointer select-none"
                      data-testid="header-size"
                    >
                      <span>
                        pci_projects_project_storages_containers_container_size_label
                      </span>
                      <span
                        class="align-middle inline-block h-4 -mt-6"
                      >
                        <ods-icon
                          class="invisible"
                          name="arrow-down"
                        />
                      </span>
                    </div>
                  </th>
                  <th
                    class="text-left pl-4 h-11 whitespace-nowrap "
                  >
                    <div
                      class="cursor-pointer select-none"
                      data-testid="header-contentType"
                    >
                      <span>
                        pci_projects_project_storages_containers_container_contentType_label
                      </span>
                      <span
                        class="align-middle inline-block h-4 -mt-6"
                      >
                        <ods-icon
                          class="invisible"
                          name="arrow-down"
                        />
                      </span>
                    </div>
                  </th>
                  <th
                    class="text-left pl-4 h-11 whitespace-nowrap "
                  >
                    <div
                      class=""
                      data-testid="header-actions"
                    >
                      <span />
                      <span
                        class="align-middle inline-block h-4 -mt-6"
                      >
                        <ods-icon
                          class="invisible"
                          name="arrow-down"
                        />
                      </span>
                    </div>
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr
                  class="border-solid border-[1px] h-[3.25rem] border-[--ods-color-blue-200]"
                >
                  <td
                    class="text-left pl-4"
                  >
                    <div>
                      <div
                        class="flex flex-col"
                      >
                        <ods-text
                          preset="span"
                        >
                          name
                        </ods-text>
                      </div>
                    </div>
                  </td>
                  <td
                    class="text-left pl-4"
                  >
                    <div
                      class=""
                    >
                      <ods-text
                        preset="span"
                      >
                        formattedDate
                      </ods-text>
                    </div>
                  </td>
                  <td
                    class="text-left pl-4"
                  >
                    <div
                      class=""
                    >
                      <ods-text
                        preset="span"
                      >
                        1 unit_size_B
                      </ods-text>
                    </div>
                  </td>
                  <td
                    class="text-left pl-4"
                  >
                    <div
                      class=""
                    >
                      <ods-text
                        preset="span"
                      >
                        contentType
                      </ods-text>
                    </div>
                  </td>
                  <td
                    class="text-left pl-4"
                  >
                    <div
                      class=""
                    >
                      <div
                        class="w-min"
                        id="navigation-action-trigger-undefined"
                      >
                        <ods-button
                          data-testid="navigation-action-trigger-action"
                          icon="ellipsis-vertical"
                          id="undefined"
                          is-disabled="false"
                          is-loading="false"
                          size="sm"
                          slot="menu-title"
                          variant="outline"
                        />
                      </div>
                      <ods-popover
                        class="py-[8px] px-0 w-max"
                        trigger-id="navigation-action-trigger-undefined"
                        with-arrow="true"
                      >
                        <div
                          class="flex flex-col"
                        >
                          <ods-button
                            class="menu-item-button w-full"
                            display-tooltip="false"
                            label="pci_projects_project_storages_containers_container_download_label"
                            size="sm"
                            variant="ghost"
                          />
                          <ods-button
                            class="menu-item-button w-full"
                            display-tooltip="false"
                            label="pci_projects_project_storages_containers_container_delete_label"
                            size="sm"
                            variant="ghost"
                          />
                        </div>
                      </ods-popover>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </ods-table>
        </div>
        <ods-pagination
          class="flex xs:justify-start md:justify-end my-8"
          default-current-page="1"
          default-items-per-page="10"
          total-items="1"
          total-pages="1"
        >
          <span
            class="mr-3"
            slot="before-total-items"
          >
            common_pagination_of
          </span>
          <span
            class="ml-3"
            slot="after-total-items"
          >
            common_pagination_results
          </span>
        </ods-pagination>
      </div>
    </div>
    <div
      class="mt-8"
    >
      <div
        class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-8 mt-16"
      >
        <a
          class="no-underline"
          href="https://docs.ovh.com/gb/en/storage/sync-rclone-object-storage/"
          target="_blank"
        >
          <ods-card
            class="w-full h-full p-[1rem]"
            color="neutral"
          >
            <div
              class="flex flex-col"
            >
              <div>
                <span
                  class="card-category text-[--ods-color-primary-500] text-[20px] leading-[28px] font-bold"
                >
                  onboarding_guide_title
                </span>
                <span
                  class="ml-[10px] card-badges-section"
                />
              </div>
              <span
                class="card-title text-[--ods-color-heading] text-[24px] leading-[32px] font-bold mb-[8px]"
              >
                pci_projects_project_storages_containers_container_documentation_title_rclone
              </span>
              <p
                class="block m-0 p-0 text-[--ods-color-text] mb-4"
              >
                pci_projects_project_storages_containers_container_documentation_description_rclone
              </p>
              <div
                class="section-see-more-label"
              >
                <ods-link
                  color="primary"
                  icon="arrow-right"
                  label="see_more_label"
                  tab-index="-1"
                />
              </div>
            </div>
          </ods-card>
        </a>
        <a
          class="no-underline"
          href="https://docs.ovh.com/gb/en/storage/s3/veeam/"
          target="_blank"
        >
          <ods-card
            class="w-full h-full p-[1rem]"
            color="neutral"
          >
            <div
              class="flex flex-col"
            >
              <div>
                <span
                  class="card-category text-[--ods-color-primary-500] text-[20px] leading-[28px] font-bold"
                >
                  onboarding_guide_title
                </span>
                <span
                  class="ml-[10px] card-badges-section"
                />
              </div>
              <span
                class="card-title text-[--ods-color-heading] text-[24px] leading-[32px] font-bold mb-[8px]"
              >
                pci_projects_project_storages_containers_container_documentation_title_veeam
              </span>
              <p
                class="block m-0 p-0 text-[--ods-color-text] mb-4"
              >
                pci_projects_project_storages_containers_container_documentation_description_veeam
              </p>
              <div
                class="section-see-more-label"
              >
                <ods-link
                  color="primary"
                  icon="arrow-right"
                  label="see_more_label"
                  tab-index="-1"
                />
              </div>
            </div>
          </ods-card>
        </a>
        <a
          class="no-underline"
          href="https://docs.ovh.com/gb/en/storage/s3/limitations/"
          target="_blank"
        >
          <ods-card
            class="w-full h-full p-[1rem]"
            color="neutral"
          >
            <div
              class="flex flex-col"
            >
              <div>
                <span
                  class="card-category text-[--ods-color-primary-500] text-[20px] leading-[28px] font-bold"
                >
                  onboarding_guide_title
                </span>
                <span
                  class="ml-[10px] card-badges-section"
                />
              </div>
              <span
                class="card-title text-[--ods-color-heading] text-[24px] leading-[32px] font-bold mb-[8px]"
              >
                pci_projects_project_storages_containers_container_documentation_title_s3-limitation
              </span>
              <p
                class="block m-0 p-0 text-[--ods-color-text] mb-4"
              >
                pci_projects_project_storages_containers_container_documentation_description_s3-limitation
              </p>
              <div
                class="section-see-more-label"
              >
                <ods-link
                  color="primary"
                  icon="arrow-right"
                  label="see_more_label"
                  tab-index="-1"
                />
              </div>
            </div>
          </ods-card>
        </a>
      </div>
    </div>
    Outlet
  </div>
</div>
`;

exports[`Show page > should display spinner 1`] = `
<div>
  <ods-spinner
    size="md"
  />
</div>
`;
